Update: ensure mouse events are on shadow root

stevematney opened the pull request:
This is to support react components that use the onMouseDown, or
onMouseUp. Currently these events are still being attached to the
document instead of the shdaow root

borland-david committed:
Update: ensure mouse events are on shadow root

This is to support react components that use the onMouseDown, or
onMouseUp. Currently these events are still being attached to the
document instead of the shdaow root

stevematney commented on src/forcedRetargeting.ts:
```diff
@@ -0,0 +1,118 @@
+      ));
+  });
+
+  return () => {
+    removeEventListeners.forEach((removeEventListener) => {
+      removeEventListener();
+    });
+  };
```
We should call this function in the `disconnectedCallback`.

borland-david replied:
I took care of this. Can you double check @stevematney

borland-david committed:
Fix: make sure to call cleanup function

stevematney commented on src/ReactHTMLElement.ts:
```diff
@@ -50,15 +55,27 @@ class ReactHTMLElement extends HTMLElement {
+    return this.retargetCleanupFunction;
+  }
+
+  set retargetCleanup(cleanupFunction: () => void) {
+    // Ensure that we cleanup an old listeners before we forget the cleanup function.
+    this.retargetCleanup();
+    this.retargetCleanupFunction = cleanupFunction;
+  }
```
We still need to run this setter when we call `retargetReactEvents`.

borland-david replied:
I thought I had done that. I guess I thought about it too much, but never did add the setters. Okay so this should work now.

stevematney commented:
It looks like this is failing to build.

borland-david committed:
Fix: remove onclick

I had accidentally left in a test to adding onClick. We do not want this
functionality as the current system for retargeting works for clicks

JosephDSchwartz commented on src/forcedRetargeting.ts:
```diff
@@ -0,0 +1,118 @@
+    }
+    el = el.parentElement;
+  }
+  return [el];
```
Since you're reassigning the param, this return will always return an array with undefined. Is that the desired result?

borland-david committed:
Fix: cleanup composedpath return

Also added node modules to be excluded from tsconfig

JosephDSchwartz commented on src/forcedRetargeting.ts:
```diff
@@ -0,0 +1,120 @@
+        || (event.composedPath && event.composedPath())
+        || composedPath(event.target);
+
+      for (let i = 0; i < path.length; i += 1) {
```
You'll get an error here I think if the `composedPath` method returns null. We might want to have an escape hatch like `(path || []).length`.

borland-david replied:
The better answer seemed to just return an empty array in composed path. That does tell me that their code would have broken in that case anyway

stevematney commented:
Since I opened this PR, I can't approve it, but with Joseph's approval (and my unofficial approval), I'm merging. 😄

aabenoja commented:
:tada: This PR is included in version 2.1.0 :tada:

The release is available on:
- [GitHub release](https://github.com/WTW-IM/react-html-element/releases/tag/v2.1.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/react-html-element/v/2.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
